### ISSUE TITLE
Introduce @CurrentSecurityContext for method arguments

### ISSUE THREAD

**Issue by jzheaux** [spring-projects/spring-security#6546]
A user can easily obtain the current user principal using the `@AuthenticationPrincipal` annotation, say in a Spring MVC application:

```java
@GetMapping("/path")
public String endpoint(@AuthenticatedPrincipal UserDetails user) {
    // ....
}
```

It would be nice to be able to do something more fundamental, like obtain the current security context:

```java
@GetMapping("/path")
public String endpoint(@CurrentSecurityContext SecurityContext context) {
    // ....
}
```
and thus:
```java
@GetMapping("/path")
public String endpoint(@CurrentSecurityContext(expression="authentication") Authentication authentication) {
    // ....
}
```

This ought to work for both Servlet and WebFlux applications, so we'd need to have two argument resolvers, one that retrieved the security context from `SecurityContextHolder` and another that got it from `ReactiveSecurityContextHolder`. Likely, they could be modeled after `AuthenticationPrincipalArgumentResolver`.

We'd also want to register these respectively in `WebMvcSecurityConfiguration` and `ServerHttpSecurityConfiguration`.

**Comment by clevertension** [comment]
@jzheaux are you working on this issue, if not, i think i can provide a PR :)

**Comment by jzheaux** [comment]
It's yours, @clevertension!

**Comment by clevertension** [comment]
do we need to support the property dig with Authentication?

```java
@GetMapping("/path")
public String endpoint(@CurrentSecurityContext(expression="authentication.details") Object details) {
    // ....
}
```

**Comment by jzheaux** [comment]
@rwinch Is there a compelling need for `errorOnInvalidType`?

**Comment by jgrandja** [comment]
Closed via #6562
